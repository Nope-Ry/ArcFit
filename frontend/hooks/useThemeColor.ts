/**
 * Learn more about light and dark modes:
 * https://docs.expo.dev/guides/color-schemes/
 */

import { useColorScheme } from 'react-native';

import { Colors } from '@/constants/Colors';

export function useThemeColor(
  props: { light?: string },
  colorName: keyof typeof Colors.light
) {
  const colorFromProps = props.light

  if (colorFromProps) {
    return colorFromProps;
  } else {
<<<<<<< HEAD
    return Colors[theme][colorName];
=======
    return Colors.light[colorName];
>>>>>>> 2e000de2698fafa3508ec8cd1bf8fc33ced3e20b
  }
}
