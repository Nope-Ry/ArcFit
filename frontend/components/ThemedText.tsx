import { Text, type TextProps, StyleSheet, Dimensions } from "react-native";

import { useThemeColor } from "@/hooks/useThemeColor";

const { width, height } = Dimensions.get("window");

export type ThemedTextProps = TextProps & {
  lightColor?: string;
  darkColor?: string;
  type?: "default" | "title" | "defaultBold" | "subtitle" | "link" | "small";
};

export function ThemedText({
  style,
  lightColor,
  darkColor,
  type = "default",
  ...rest
}: ThemedTextProps) {
  const color = useThemeColor({ light: lightColor, dark: darkColor }, "text");

  return (
    <Text
      style={[
        { color },
        type === "default" ? styles.default : undefined,
        type === "title" ? styles.title : undefined,
        type === "defaultBold" ? styles.defaultBold : undefined,
        type === "subtitle" ? styles.subtitle : undefined,
        type === "link" ? styles.link : undefined,
        type === "small" ? styles.small : undefined,
        style,
      ]}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  default: {
    fontSize: width * 0.04, // Example: 4% of the screen width
    lineHeight: height * 0.03, // Example: 3% of the screen height
  },
  defaultBold: {
    fontSize: width * 0.04, // Example: 4% of the screen width
    lineHeight: height * 0.03, // Example: 3% of the screen height
    fontWeight: '600',
  },
  title: {
    fontSize: width * 0.08, // Example: 8% of the screen width
    fontWeight: 'bold',
    lineHeight: height * 0.04, // Example: 4% of the screen height
  },
  subtitle: {
    fontSize: width * 0.05, // Example: 5% of the screen width
    fontWeight: 'bold',
  },
  small: {
    fontSize: width * 0.03, // Example: 3% of the screen width
    lineHeight: height * 0.02, // Example: 2% of the screen height
  },
  link: {
    lineHeight: height * 0.04, // Example: 4% of the screen height
    fontSize: width * 0.04, // Example: 4% of the screen width
    color: '#0a7ea4',
  },
});