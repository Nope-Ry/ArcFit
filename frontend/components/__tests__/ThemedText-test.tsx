import React from "react";
import { render } from "@testing-library/react-native";
import { ThemedText } from "../ThemedText"; // 替换为实际路径
import { useThemeColor } from "@/hooks/useThemeColor";

jest.mock("@/hooks/useThemeColor", () => ({
  useThemeColor: jest.fn(),
}));

describe("ThemedText Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders with default type and theme color", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#000");

    const { getByText } = render(
      <ThemedText lightColor="#fff">Default Text</ThemedText>
    );

    const textElement = getByText("Default Text");

    // 验证文本内容
    expect(textElement).toBeTruthy();

    // 验证样式是否包含默认样式
    expect(textElement.props.style).toContainEqual({ color: "#000" });
    expect(textElement.props.style).toContainEqual({
      fontSize: expect.any(Number),
      lineHeight: expect.any(Number),
    });
  });

  it("renders with 'title' type and correct styles", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#0a7ea4");

    const { getByText } = render(
      <ThemedText type="title">Title Text</ThemedText>
    );

    const textElement = getByText("Title Text");

    // 验证文本内容
    expect(textElement).toBeTruthy();

    // 验证样式是否包含 'title' 类型的样式
    expect(textElement.props.style).toContainEqual({ color: "#0a7ea4" });
    expect(textElement.props.style).toContainEqual({
      fontSize: expect.any(Number),
      fontWeight: "bold",
      lineHeight: expect.any(Number),
    });
  });

  it("renders with 'link' type and correct styles", () => {
    const { getByText } = render(<ThemedText type="link">Link Text</ThemedText>);

    const textElement = getByText("Link Text");

    // 验证文本内容
    expect(textElement).toBeTruthy();

    // 验证样式是否包含 'link' 类型的样式
    expect(textElement.props.style).toContainEqual({
      fontSize: expect.any(Number),
      lineHeight: expect.any(Number),
      color: "#0a7ea4",
    });
  });

  it("applies custom styles passed via the `style` prop", () => {
    const { getByText } = render(
      <ThemedText style={{ fontSize: 20, color: "#ff0000" }}>
        Custom Styled Text
      </ThemedText>
    );

    const textElement = getByText("Custom Styled Text");

    // 验证文本内容
    expect(textElement).toBeTruthy();

    // 验证自定义样式是否生效
    expect(textElement.props.style).toContainEqual({
      fontSize: 20,
      color: "#ff0000",
    });
  });
});