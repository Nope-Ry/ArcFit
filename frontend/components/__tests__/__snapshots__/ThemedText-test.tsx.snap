// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  style={
    [
      {
        "color": "#11181C",
      },
      {
        "fontSize": 30,
        "lineHeight": 40.019999999999996,
      },
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
    ]
  }
>
  Snapshot test!
</Text>
`;
